### PR TITLE
Improve `contains` operator

### PR DESCRIPTION
**Pull request by yaisog**
Currently, the `contains` operator checks the specified field of each input tiddler consecutively for the parameter, resulting in O(n) performance.
With this PR, we will call `wiki.findListingsOfTiddler()` to get a (cached) list of tiddlers that have the parameter in their respective field and then intersect this intermediate list with the input list, resulting in O(1) performance.

This uses ES2017 and would be appropriate for v5.4.0. The change is fully backward compatible and could be released in an earlier version, but would need to be rewritten for that purpose.

The PR was ultimately amended to make the search method choosable and to allow for title lists to be passed as parameter of which either all or any must match, also choosable. For this behavior to be more flexible, the field suffix and the new flags were moved into additional parameters to the operator, which, unlike suffixes, can be variables or transclusions.

### Performance Comparison
`[all[tiddlers]count[]]` → 16319

#### Methodology
I used `$tw.wiki.filterTiddlers(testFilter)` in a console script to run various filters. The steps were:
1. Run the filter to possibly initially the global cache
2. Run the filter again and measure runtime using `performance.now()`
3. Set the `list` field of a tiddler with `$tw.wiki.setText()`
4. Run the filter again and measure runtime (now includes cache rebuild)
5. Run the filter again and measure runtime (should be similar to the result from step 2)

The script repeated these 5 steps 100 times for each filter and computed the median runtime.
I was lazy and had an LLM create most of this script and only verified it.

#### Results
`[all[tiddlers]first[10]]` - baseline without `contains`
Time2 - 0.40, Time4 - 0.40, Time5 - 0.40

All times are in ms
All below filters return 0 results, so that the output list length does not affect timing.

`[all[tiddlers]first[10]contains[Test]]` - `contains` is the v5.3.8 implementation
Time2 - 0.40, Time4 - 0.40, Time5 - 0.40
As expected, there is no penalty for cache rebuild
The results are identical to the baseline results, i.e. we have reached the time measurement resolution limit.

`[all[tiddlers]contains[Test]]`
Time2 - 3.20, Time4 - 3.20, Time5 - 3.10
This is significantly longer than the baseline value

`[all[tiddlers]first[10]contains2[Test]]` - `contains2` is this PR (inital commit, before adding any additional suffixes / parameters)
Time2 - 0.60, Time4 - 2.80, Time5 - 0.50
The penalty for cache rebuilding seems to be around 2 ms

`[all[tiddlers]contains2[Test]]`
Time2 - 0.40, Time4 - 2.40, Time5 - 0.40
Without cache rebuild, this corresponds to the baseline value.
The number of input tiddlers at which the new algorithm becomes faster for a single call seems to be around 9 000. If the filter is run multiple times during a refresh, the then-cached execution is much faster for any subsequent runs.
Also, in a smaller wiki the cache rebuild will be faster, so the break-even point is at a lower tiddler count.

`[all[tiddlers]contains[Test],[],[]]` - this PR as of 2025-10-24 (with the additional options)
Time2 - 4.20, Time4 - 4.30, Time5 - 4.20
This is about a ms slower than the v5.3.8 implementation. I am not quite sure why.

`[all[tiddlers]contains[Test],[],[lookup]]`
Time2 - 0.60, Time4 - 2.30, Time5 - 0.60
About 0.2ms slower than `contains2`

#### Note
In the wiki that prompted this change, the `contains` operator is used up to some tens of times, so the milliseconds can add up quickly. In the worst case (single use) it will add 2 ms cache rebuild penalty, but in other cases it can save tens of ms of refresh time.

### CODE REVIEW

File: core/modules/filters/contains.js [hunk 1]
```diff
@@ -13,28 +13,15 @@ Filter operator for finding values in array fields
...
-		source(function(tiddler,title) {
-			if(tiddler) {
-				var list = tiddler.getFieldList(fieldname);
-				if(list.indexOf(operator.operand) !== -1) {
-					results.push(title);
-				}
-			}
-		});
-	}
+	const results = [];
```
Comment (pmario) [inline core/modules/filters/contains.js:15]: imo it should be `let results`

File: core/modules/filters/contains.js [hunk 2]
```diff
@@ -13,28 +13,15 @@ Filter operator for finding values in array fields
...
+	const fieldname = operator.suffix || "list";
+	const tiddlersWithOperand = options.wiki.findListingsOfTiddler(operator.operand, fieldname);
```
Comment (pmario) [inline core/modules/filters/contains.js]: `.findListingsOfTiddler` uses the `this.getGlobalCache()`, which is constantly destroyed with every wiki store change. That's basically with every keypress and mouse event. 

Did you do any performance measurements, if the new function is really faster as the old one. Because, if the cache has to be rebuilt everytime the filter is used, it will be much slower than the existing functionality. 

You need to test this function with a very large number of tiddlers, because building the globalchache iterates over all existing tiddlers and not only the input titles.

Comment (pmario) [reply core/modules/filters/contains.js]: Also see: https://github.com/TiddlyWiki/TiddlyWiki5/issues/4572

Comment (saqimtiaz) [reply core/modules/filters/contains.js]: The current implementation is by design and is optimized for a smaller number of input tiddlers to the operator, compared to the total number of tiddlers in the wiki.

File: core/modules/filters/contains.js (hunk 1 above)
Comment (yaisog) [reply core/modules/filters/contains.js:15]: I am not reassigning `results`, but only mutating the array it points to. `const` communicates this intent more clearly.

File: core/modules/filters/contains.js (hunk 2 above)
Comment (yaisog) [reply core/modules/filters/contains.js]: > The current implementation is by design and is optimized for a smaller number of input tiddlers to the operator, compared to the total number of tiddlers in the wiki.

That is not always _my_ usecase and it is documented nowhere. In a large wiki with some tens of input tiddlers it makes a huge difference of almost 100 ms total time. Optimizing for runtime when runtime is very small anyway – with few input tiddlers – is probably not the best UX choice. At least not in my opinion, but I am regularly proven wrong.

Comment (yaisog) [reply core/modules/filters/contains.js]: > Did you do any performance measurements, if the new function is really faster as the old one.

I compared these two filters:
`[[TiddlerA]contains[TiddlerB]]`
where TiddlerA has a list field with 10 entries to
`[[TiddlerB]listed[]match[TiddlerA]]`
since `listed[]` uses the same cache.
Using the _Advanced Performance_ plugin, I got the timings back as 3.5 ms vs. 0.1 ms (average of multiple runs), so even for a single input tiddler this is much faster.
I put these filters in a tiddler, so that tiddler changed right before running the filters. Where the cache update fits into this process, I can't tell. I tested this in a wiki with nearly 12 000 tiddlers.

Comment (yaisog) [reply core/modules/filters/contains.js]: > Using the _Advanced Performance_ plugin, I got the timings back as 3.5 ms vs. 0.1 ms (average of multiple runs), so even for a single input tiddler this is much faster.

@pmario @saqimtiaz I have to backtrack on this. This was inadvertently done in a wiki where I had already replaced the `contains` operator with the one from this PR. So, the 3.5 ms timing is for the _new_ operator, and `listed` was so quick likely because the cache had just been built for `contains` (which came just before it in the wikitext). The original core `contains` is < 0.1 ms in this case.

The argument still stands, though, that an optimization that saves 3 ms in a design use case, but adds nearly 100 ms in other valid use cases is not much of an optimization. We may choose to keep it that way, though, and I'll maybe make a plugin instead. Then we should add a note to the docs about this optimization, though, maybe with a link to `listed[]`. Unfortunately, the core doesn't have a filter to check if input tiddlers are contained in a parameter list; that would be of great help when doing it the `listed` way as shown above.

**Comment by saqimtiaz** [comment]
@yaisog the current implementation favours optimizing for wikis with a large number of tiddlers compared to the input to the operator. This is partly because while it is possible to use selection constructors to narrow down an input list to an operator, there are no possiblities to reduce the total number of tiddlers in a wiki.

A good approach here might be to allow the alternative implementation via an operator suffix, and document the strengths of the two modes.

**Comment by yaisog** [comment]
@saqimtiaz I like that idea. I was initially going to propose to put in condition in the code that chose one way or the other, based on number of total tiddlers and/or input tiddlers. But it's very labor-intensive to make this truly optimal. A suffix and appropriate documentation would allow users to choose their own optimum.

There are two downsides:
- We'd have double suffixes like `contains:field:listed[]`
- Suffixes are generally too inflexible because they cannot be soft parameters. For `contains` in particular, I'd like to move the `field` suffix to a second, optional parameter. The method indicator _could_ be a suffix, but what if I want to make the filter operate differently for small or large input lists with pure wikitext? Then it would be nice to have the method be a parameter, too. That's actually a discussion that is valid for many filter operators that have suffixes because in ye olden days we didn't have the mechanism of multiple parameters. For the sake of backwards compatibility we could have both, though it would bloat the code a bit.

Before I start writing, I'd like to hear @Jermolene's thoughts on this, though.

**Comment by saqimtiaz** [comment]
We will likely need a more intuitive name than `listed` for the alternative mode. 

The use of an extra parameter is appealing for the added flexibility. However, we need to be sure that we aren't precluding the ability to use multiple parameters for other purposes, such as searching for multiple terms, for example:
- `contains:any<TermOne>,<TermTwo>`
- `contains:all<TermOne>,<TermTwo>`

In the long term we need to find a more flexible affordance for assigning suffixes using variables.

**Comment by yaisog** [comment]
Hi @saqimtiaz, you can only get so far with positional parameters/operands. Maybe there'll be named filter parameters someday...

`any` and `all` is something I would like to also see settable via variables. In any case with more than one term, I'd rather see the first parameter be a title list of terms than have a large number of parameters that are all filled with different variables. The third parameter could then be a `flags` parameter, e.g. `[all,listed]` with a default of `[any]`. Though that is less flexible or at least more awkward to set via a variable, but still better than a comma-separated suffix.
I am using `listed` as a placeholder until we'll have a consensus after an invariably lengthy discussion. 😜

**Comment by yaisog** [comment]
Alrighty, since it seemed useful to me I went ahead and implemented the previously discussed items.
The syntax now would be as follows:
```
contains[titlelist],[field],[flags]
```

- `titlelist` can be a single literal title or a soft parameter containing a title list. How a list is handled, depends on the `flags` parameter, see below.
- `field` determines the field of the input tiddlers that contains the list to be searched. If empty `[]` it defaults to `list`.
- `flags` is a comma-separated list of modifiers:
  - `some` will return all input titles whose list contains at least one title from the parameter list. If not given, only those input titles whose list contains _all_ titles from the parameter list are returned - this corresponds to the default `words` mode of the `search` operator. Only useful when the `titlelist` parameter contains a list of at least two titles.
  - `listed` will use the alternative cache lookup method initially proposed. If not given, the legacy algorithm is used.

Examples:
```
contains[TiddlerA]
contains:tags[TiddlerA]

contains[TiddlerA],[],[listed]

<$let titleList="TiddlerA [[Tiddler B]] TiddlerC">
  <$list filter="[contains<titleList>,[tags],[some,listed]]" />
</$let>
```
The first two examples show legacy usage which still works, using the legacy algorithm.
The third example demonstrates the empty `field` parameter.
The fourth example shows how to check for multiple items.

**Comment by pmario** [comment]
IMO the first parameter of contains is the "search value". [See docs](https://deploy-preview-9351--tiddlywiki-previews.netlify.app/#contains%20Operator). 

```
contains[search-term],[field],[flags]
```

**Comment by pmario** [comment]
There seems to be a bug. To reproduce: 

- create a tiddler: **asdf** with `list: aaa bbb ccc ddd eee`
- Test this filter in Advanced search `[[asdf]contains[]]` -> Bug. It returns "asdf"

image

**Comment by yaisog** [comment]
> IMO the first parameter of contains is the "search value".

Right, that's how it is implemented.

> There seems to be a bug.

Correct. An empty parameter currently does not return anything. Technically, it could be argued if that should really be the case, since the list in `asdf` also contains the empty item. The docs don't explicitly mention that. But we should stick with the current behavior.

**Comment by pmario** [comment]
>But we should stick with the current behavior.

What do you mean by current behaviour? v5.3.8 has a different behaviour, than this PR. IMO we need to keep v5.3.8 behaviour, otherwise it would be a regression.

**Comment by yaisog** [comment]
@pmario _current_ = currently released version = v5.3.8

Since this PR targets v5.4, I think we can at least think about using a different treatment for edge cases, if such a treatment be sensible.

What would be the natural answer to the question whether some list contained an empty string?
_Currently_, there is an inconsistency with the `search` operator. That operator returns all input titles when the search string is empty, unlike `contains`. I'd rather make TW more consistent than to force backwards compatibility here.

Interestingly, the current PR state returns nothing when used with the `some` flag, so there is inherent inconsistency that must be addressed in any case.

**Comment by yaisog** [comment]
Since we cannot know if a search parameter of "TiddlerA TiddlerB" is a title or a list, I have added an additional flag `list` to indicate that the parameter is a list and treat it as such.
To avoid excessive confusion, I also renamed the `listed` flag to `lookup`, though still wide open for debate.
The example above would now be
```
<$let titleList="TiddlerA [[Tiddler B]] TiddlerC">
  <$list filter="[contains<titleList>,[tags],[some,lookup,list]]" />
</$let>
```

**Comment by pmario** [comment]
I think this PR mixes 2 completely different topics. It should be split into 2 PRs. 

1. performance optimisation, which will need some tests, because I think `globalChache` does not really help
2. Empty parameter eg: `a b c +[contains[]]` will result in `a b c` as output

Add 2) I do have strong concerns here. IMO contains[] and search[] can not be compared because they are different things in set-theory. IMO they do not need to be consistent. I need time for more research.

Add 1) I do have no problem if performance can be improved. ... If it really is.

**Comment by pmario** [comment]
Test this `"" +[contains[]] ` in AdvancedSearch Fiter tab -> **0 matches**

Then test this: `[[]contains[]]` -> **1 matches but no output**  --> This will cause problems, which in turn will make filter code more complex, to catch this edge cases. 

`[<var>] +[contains[]]` -> 1 matches

**Comment by yaisog** [comment]
> Test this `"" +[contains[]] ` in AdvancedSearch Fiter tab -> **0 matches**
> 
> Then test this: `[[]contains[]]` -> **1 matches but no output**

You get the same result if you omit the `contains` operator in both cases. A filter operator cannot fix TW parsing idiosyncrasies.

**Comment by yaisog** [comment]
> I think this PR mixes 2 completely different topics.

I dare to offer more:
3. Enable lists to be passed as search parameter to check for more than one item at a time
4. Move from suffix parametrization to comma-separated parameters
5. Allow for behavior tuning via multiple flags

All of these could probably go into separate PRs. However, it is helpful for neither the author nor possible reviewers to launch multiple PRs simultaneously for a single source file.

**Comment by yaisog** [comment]
> contains[] and search[] can not be compared because they are different things

As a non-mathematician, probably like most users, I see a close connection between a `search` operator that checks if a substring is contained in a string and a `contains` operator that checks if an item is contained in a list. I believe most users would expect these two to behave similarly, even if set theory says they're not.

**Comment by saqimtiaz** [comment]
I think it would help a great deal if the proposed spec was presented an issue first, to facilitate discussion, without getting caught up in the intricacies of the implementation. At present I don't think we have a specification for the proposed implementation written down in one place, which makes a review more difficult.
 
Two of the bigger picture questions that I think need answering are:
- can we find a way to allow variables to be used for filter run suffixes? The alternative is to rewrite all operators that use suffixes.
- what our default should be for passing multiple titles to an operator: multiple parameters, a single parameter as a titlelist, or perhaps a single [multivalued variable](https://github.com/TiddlyWiki/TiddlyWiki5/pull/8972) as a parameter?

**Comment by yaisog** [comment]
Hi @saqimtiaz, as you know this PR was started without creating an issue first, because initially it was just a performance improvement without any interface change.

Considering your bigger picture questions, don't you think that there will always be another idea or project on the horizon that would need to be resolved for fear of creating a backward-compatibility nightmare? MVVs are dormant for more than half a year. Variable suffixes are being discussed since suffixes were introduced. I'd love to see named filter parameters and might want to wait for these to appear.

I don't think that creating a GitHub issue for discussion or putting any more thought into `contains` - or maybe any other improvement aside from minor docs corrections or bug fixes - makes sense in such a situation.

I was going to rudimentarily extend the docs page for this `contains` proposal to ease testing, but you successfully convinced me that this is not the right time. Maybe I'll wait out the formerly lively discussion on Talk TW about how we should color or format our docs examples to not have to do this twice.

In the meantime, I'll do what I usually do and resolutely plan to make a plugin, which more often than not never materializes. It was a fun exercise, though, and I'm saving a lot of refresh time in my own wikis with this modification. The title list parameter is just a cherry on top that I'll probably never need.

I'll leave the PR open. Please ping me at any point in the future when you feel that the big picture is sufficiently clear to move forward. I'll be happy to pick this up again.

**Comment by saqimtiaz** [comment]
PRs quite often end up being more involved and far reaching than originally planned. However, it is then good practice to update the PR description with a current summary of the proposed changes, which would suffice just as well as creating an issue. As mentioned above, what is needed to facilitate review is a cohesive specification for the proposed implementation written down in one place.

Multi valued variables are on the roadmap for v5.4.0 - which unless something drastic occurs will be the next TiddlyWiki release. Therefore it is on the same timeline as this proposal. As regards to the use of variables in suffixes, we are well placed to make a policy decision as to whether we want to pursue this and whether its even feasible, or whether updating the operators to use additional parameters instead of existing suffixes is the way to go. 

However, making any progress on these fronts requires discussion, without which the proposed improvements will likely linger in limbo until someone champions it again. Discussing the merits of each proposed change, especially in the broader context of other changes being introduced and other needs that need addressing, is a critical part of how we do core development and your participation in these discussions would be most welcome as I find your work very promising.

**Comment by yaisog** [comment]
Since MVVs are coming with v5.4, I'm favoring using these instead of the `list` flag, i.e. letting the search parameter potentially be an MVV and handle it accordingly to the `words` or `some` flags.
Will `operands[0]` be an array in this case or how else is it implemented? Is there a canonical way to handle these, yet?

**Comment by yaisog** [comment]
I'm tagging #8972 and #9364 because of their relevance to this PR.

**Comment by yaisog** [comment]
Closing this since the performance gains are only circumstantial and the additional functionality was implemented via the `intersection` operator in #9487.
